control_plane/src/background_process.rs

```diff
@@ -0,0 +1,261 @@
+//! invocations and cannot possibly know about particular running processes otherwise.
+//!
+//! See [`lock_file`] module for more info.
+
+use std::ffi::OsStr;
+use std::io::Write;
+use std::path::Path;
+use std::process::{Child, Command};
+use std::time::Duration;
+use std::{fs, io, thread};
+
+use anyhow::{anyhow, bail, Context, Result};
+use nix::errno::Errno;
+use nix::sys::signal::{kill, Signal};
+use nix::unistd::Pid;
+
+use utils::lock_file;
+
+const RETRIES: u32 = 15;
+const RETRY_TIMEOUT_MILLIS: u64 = 500;
+
+/// Ensures process startup handles pidfile, either by creating the new one
+/// or by checking the one that the spawned process should create.
+pub enum InitialPidFile<'t> {
+    /// Creates a special pidfile, to allow future CLI invocations to manipulate the process.
+    Create(&'t Path),
+    /// A process is creating the pidfile itself, need to wait for that event.
+    Expect(&'t Path),
+}
+
+/// Starts the background child process using the parameters given.
+/// Ensure that the pid file is created and contains the correct pid value.
+pub fn start_process<F, S: AsRef<OsStr>>(
+    process_name: &str,
+    datadir: &Path,
+    command: &Path,
+    args: &[S],
+    initial_pid_file: InitialPidFile,
+    process_status_check: F,
+) -> anyhow::Result<Child>
+where
+    F: Fn() -> anyhow::Result<bool>,
+{
+    let log_path = datadir.join(format!("{process_name}.log"));
+    let process_log_file = fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .append(true)
+        .open(&log_path)
+        .with_context(|| {
+            format!("Could not open {process_name} log file {log_path:?} for writing")
+        })?;
+    let same_file_for_stderr = process_log_file.try_clone().with_context(|| {
+        format!("Could not reuse {process_name} log file {log_path:?} for writing stderr")
+    })?;
+
+    let mut command = Command::new(command);
+    let background_command = command
+        .stdout(process_log_file)
+        .stderr(same_file_for_stderr)
+        .args(args);
+    let filled_cmd = fill_aws_secrets_vars(fill_rust_env_vars(background_command));
+
+    let mut spawned_process = filled_cmd.spawn().with_context(|| {
+        format!("Could not spawn {process_name}, see console output and log files for details.")
+    })?;
+    let pid = spawned_process.id();
+    let pid = Pid::from_raw(
+        i32::try_from(pid)
+            .with_context(|| format!("Subprocess {process_name} has invalid pid {pid}"))?,
+    );
+
+    let pid_file_to_check = match initial_pid_file {
+        InitialPidFile::Create(target_pid_file_path) => {
+            match lock_file::create_lock_file(target_pid_file_path, pid.to_string()) {
+                lock_file::LockCreationResult::Created { .. } => {
+                    // drop the file that's created, hence drop the lock: we cannot hold it forever since the cli will exit anyway
```

Comment:
My point here is not to protect, but to create a pid file that we could later reuse to kill etcd.
Etcd does create its own lockfile, so bothering with wrapping it makes no sense, but we still somehow need to know the pid.

I've reused the same write with lock method, since it's already present, but could rewrite it into plain file creation, if it's confusing.

Reply:
I would turn it into usual write or at least comment that we needs this just for etcd. Up to you